### PR TITLE
Alm 1008 - ruby 2.6.5

### PR DESCRIPTION
**Pull request by sgatchell**
https://jira.plos.org/jira/browse/ALM-1008

This upgrades ruby to 2.6.5, which requires an upgrade to bundler which in turn results in errors with the mysql2 and other gems.  I chose to fix it by updating rails.  I found that slowly incrementing gems was slow so I went the other route and removed restrictions on gems besides `rails` and pared others back as needed.

To test, follow the instructions in the README, particularly the docker instructions.

### CODE REVIEW

**Comment by egh** [comment]
@sgatchell It looks like this broke the test coverage reporting in teamcity?

**Comment by sgatchell** [comment]
Oops looks like it did.  Thanks @egh

**Comment by egh** [comment]
It would be nice to have a `.ruby-version` file.

**Comment by sgatchell** [comment]
done! `echo 2.6.5 > .ruby-version`

**Comment by sgatchell** [comment]
Regarding the code coverage failure, it looks like it's ok on the dockerized build:  https://teamcity.plos.org/teamcity/viewType.html?buildTypeId=Alm_Builds_CodeCoverageUnitTests_LagottoBuildNewDockerBased&branch_Alm_Builds_CodeCoverageUnitTests=pull%2F51&tab=buildTypeStatusDiv

It's broken on the non-dockerized build because the agent doesn't have ruby 2.6.5 installed.  I'm going to disable the non-dockerized build if no objection: https://teamcity.plos.org/teamcity/viewType.html?buildTypeId=Alm_Builds_CodeCoverageUnitTests_LagottoBuild&branch_Alm_Builds_CodeCoverageUnitTests=pull%2F51&tab=buildTypeStatusDiv

File: Gemfile [hunk 1]
```diff
@@ -1,105 +1,106 @@
...
+gem "github-markdown"
+gem "rouge"
+gem 'dotiw'
 
-gem 'sprockets-rails', '~> 2.2.0', :require => 'sprockets/railtie'
-gem 'sass-rails', '~> 4.0.4'
-gem "uglifier", "~> 2.5.3"
-gem 'coffee-rails', '~> 4.1.0'
+gem 'sprockets-rails', require: 'sprockets/railtie'
+gem 'sass-rails', '5.0.7' # breaking change in 6.0
```
Comment (egh) [inline Gemfile]: I think it might be better to use `~> 5.0.7` to allow for upgrades in `5.0.x` while preventing an upgrade to 6.

File: Gemfile [hunk 2]
```diff
@@ -1,105 +1,106 @@
...
+  gem "factory_girl_rails", require: false
+  gem "capybara"
+  gem 'capybara-screenshot'
+  gem 'colorize'
+  gem "database_cleaner", '1.6.2' # before they added safeguards
+  gem "launchy"
+  gem "email_spec"
+  gem "rack-test", require: 'rack/test'
+  gem "simplecov", require: false
+  gem 'codeclimate-test-reporter', '0.4.8', require: false #simplecov still finds test cases
```
Comment (egh) [inline Gemfile]: I don't think we use codeclimate, maybe we can drop this?

File: Gemfile [hunk 3]
```diff
@@ -1,105 +1,106 @@
...
 group :test, :development do
   gem 'byebug'
-  gem "rspec-rails", "~> 3.1.0"
-  # gem 'spring-commands-rspec', '~> 1.0.4'
-  gem 'teaspoon-jasmine', '~> 2.2.0'
-  gem "brakeman", "~> 2.6.0", :require => false
-  gem 'rubocop', '~> 0.49.0'
-  gem 'bullet', '~> 4.14.0'
+  gem "rspec-rails"
+  # gem 'spring-commands-rspec'
```
Comment (egh) [inline Gemfile]: Can we just delete this line?

File: Gemfile [hunk 4]
```diff
@@ -1,105 +1,106 @@
...
-  # gem 'spring', '~> 1.1.2'
-  gem 'hologram', '~> 1.3.1'
+  gem 'pry-rails'
+  gem 'better_errors'
+  gem 'binding_of_caller'
+  gem 'capistrano'
+  gem 'capistrano-rails', require: false
+  gem 'capistrano-bundler', require: false
+  gem 'capistrano-npm'
+  # gem 'spring'
```
Comment (egh) [inline Gemfile]: Can we just delete this line?

File: config/spring.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+%w(
```
Comment (egh) [inline config/spring.rb]: If we delete the `spring` lines in `Gemfile`, maybe we can delete this file? Or are you using spring locally?

**Comment by egh** [comment]
Checked this out locally & ran the docker tests, but i get an error:

```
$ docker-compose run --rm appserver rspec
...
An error occurred in a `before(:suite)` hook.
Failure/Error: DatabaseCleaner.clean_with(:truncation)

ActiveRecord::NoDatabaseError:
  Unknown database 'alm'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_adapters/mysql2_adapter.rb:29:in `rescue in mysql2_connection'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_adapters/mysql2_adapter.rb:11:in `mysql2_connection'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:729:in `new_connection'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:773:in `checkout_new_connection'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:752:in `try_to_checkout_new_connection'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:713:in `acquire_connection'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:490:in `checkout'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:364:in `connection'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:883:in `retrieve_connection'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_handling.rb:128:in `retrieve_connection'
# /usr/local/bundle/gems/activerecord-5.0.7.2/lib/active_record/connection_handling.rb:91:in `connection'
# /usr/local/bundle/gems/database_cleaner-1.6.2/lib/database_cleaner/active_record/truncation.rb:234:in `clean'
# /usr/local/bundle/gems/database_cleaner-1.6.2/lib/database_cleaner/base.rb:46:in `clean_with'
# /usr/local/bundle/gems/database_cleaner-1.6.2/lib/database_cleaner/configuration.rb:91:in `block in clean_with'
# /usr/local/bundle/gems/database_cleaner-1.6.2/lib/database_cleaner/configuration.rb:91:in `each'
# /usr/local/bundle/gems/database_cleaner-1.6.2/lib/database_cleaner/configuration.rb:91:in `clean_with'
# ./spec/rails_helper.rb:143:in `block (2 levels) in <top (required)>'
# ------------------
# --- Caused by: ---
# Mysql2::Error:
#   Unknown database 'alm'
#   /usr/local/bundle/gems/mysql2-0.5.3/lib/mysql2/client.rb:90:in `connect'
```

**Comment by sgatchell** [comment]
I get a slightly different error when I skip the seeding step.  I suspect this means the image has not been built since this branch was pulled. Executing all the steps [here](https://github.com/PLOS/lagotto#running-the-tests) (down, build, seed, test) I see passing tests.

**Comment by egh** [comment]
@sgatchell ugh, sorry. I should have read the README more carefully.

File: config/spring.rb (hunk 5 above)
Comment (sgatchell) [reply config/spring.rb]: @egh I'd rather not invest in polishing this old turd of an app.

However, if you are confident that it's harmless to remove spring (I'm semi-confident but don't want to spend more time) then I'd be glad to remove the spring init code that I carried forward in a number of rake tasks. It's one less deviance from the config generated by the rails application updater.

Comment (egh) [reply config/spring.rb]: lol, I hear you. I just thought it was commented out in `Gemfile*` so I thought it wasn't even being used?

Comment (sgatchell) [reply config/spring.rb]: True that.  And, now that tests are passing I can be more confident about making changes.  I'll remove the spring stuff, using a small timebox.

File: config/initializers/new_framework_defaults.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Be sure to restart your server when you modify this file.
```
Comment (egh) [inline config/initializers/new_framework_defaults.rb]: Do we want to flip these? Or is it not worth the bother.

File: config/environments/production.rb [hunk 7]
```diff
@@ -26,33 +51,49 @@
...
   config.action_mailer.default_url_options = { :host => "#{ENV['MAIL_ADDRESS']}:#{ENV['MAIL_PORT']}" }
+
+  # Use default logging formatter so that PID and timestamp are not suppressed.
+  config.log_formatter = ::Logger::Formatter.new
+
+  # Use a different logger for distributed setups.
+  # require 'syslog/logger'
+  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new 'app-name')
+
+  if ENV["RAILS_LOG_TO_STDOUT"].present?
```
Comment (egh) [inline config/environments/production.rb]: We always want to log to stdout, no? Maybe we should just remove this and make it always log to stdout.

File: config/initializers/new_framework_defaults.rb (hunk 6 above)
Comment (sgatchell) [reply config/initializers/new_framework_defaults.rb]: Not worth the verification. Some of these changes are subtle, like changing whether returning false breaks the `before_action` callback chain.  

The current values represent older rails conventions which is what lagotto expects.

File: config/environments/production.rb (hunk 7 above)
Comment (sgatchell) [reply config/environments/production.rb]: Agreed! Good call.

**Comment by egh** [comment]
> I get a slightly different error when I skip the seeding step. I suspect this means the image has not been built since this branch was pulled. Executing all the steps here (down, build, seed, test) I see passing tests.

Thanks, @sgatchell , that fixed it.

**Pull request merged**